promql/info.go

```diff
@@ -0,0 +1,467 @@
+				ss = seriesAndTimestamp{Series{Metric: sample.Metric}, ts}
+			}
+			addToSeries(&ss.Series, enh.Ts, sample.F, sample.H, numSteps)
+			seriess[h] = ss
+		}
+	}
+
+	// Reuse the original point slices.
+	for _, m := range origMatrices {
+		for _, s := range m {
+			putFPointSlice(s.Floats)
+			putHPointSlice(s.Histograms)
+		}
+	}
+	// Assemble the output matrix. By the time we get here we know we don't have too many samples.
+	numSamples := 0
+	output := make(Matrix, 0, len(seriess))
+	for _, ss := range seriess {
+		numSamples += len(ss.Floats) + totalHPointSize(ss.Histograms)
+		output = append(output, ss.Series)
+	}
+	ev.currentSamples = originalNumSamples + numSamples
+	ev.samplesStats.UpdatePeak(ev.currentSamples)
+	return output, warnings
+}
+
+func (ev *evaluator) gatherVector(ts int64, input Matrix, output Vector) Vector {
+	output = output[:0]
+	for i, series := range input {
+		switch {
+		case len(series.Floats) > 0 && series.Floats[0].T == ts:
+			output = append(output, Sample{Metric: series.Metric, F: series.Floats[0].F, T: ts, OrigT: series.Floats[0].OrigT})
+			// Move input vectors forward so we don't have to re-scan the same
+			// past points at the next step.
+			input[i].Floats = series.Floats[1:]
+		case len(series.Histograms) > 0 && series.Histograms[0].T == ts:
+			output = append(output, Sample{Metric: series.Metric, H: series.Histograms[0].H, T: ts, OrigT: series.Histograms[0].OrigT})
+			input[i].Histograms = series.Histograms[1:]
+		default:
+			continue
+		}
+
+		// Don't add histogram size here because we only
+		// copy the pointer above, not the whole
+		// histogram.
+		ev.currentSamples++
+		if ev.currentSamples > ev.maxSamples {
+			ev.error(ErrTooManySamples(env))
+		}
+	}
+	ev.samplesStats.UpdatePeak(ev.currentSamples)
+
+	return output
+}
+
+// combineWithInfoVector combines base and info Vectors.
+// Base series in ignoreSeries are not combined.
+func (ev *evaluator) combineWithInfoVector(base, info Vector, ignoreSeries map[int]struct{}, baseSigs []map[string]string, infoSigs []string, enh *EvalNodeHelper, dataLabelMatchers map[string][]*labels.Matcher) (Vector, error) {
+	if len(base) == 0 {
+		return nil, nil // Short-circuit: nothing is going to match.
+	}
+
+	// All samples from the info Vectors hashed by the matching label/values.
+	if enh.infoSamplesBySig == nil {
+		enh.infoSamplesBySig = make(map[string]Sample, len(enh.Out))
+	} else {
+		clear(enh.infoSamplesBySig)
+	}
+
+	for i, s := range info {
+		sig := infoSigs[i]
+		if existing, exists := enh.infoSamplesBySig[sig]; exists {
+			switch {
+			case existing.OrigT > s.OrigT:
+				// Keep the other info sample, since it's newer.
+			case existing.OrigT < s.OrigT:
+				// Keep this info sample, since it's newer.
+				enh.infoSamplesBySig[sig] = s
+			default:
+				// The two info samples have the same timestamp - conflict.
+				name := s.Metric.Map()[labels.MetricName]
+				ev.errorf("found duplicate series for info metric %s", name)
```

Comment:
I think we could do this even if they collide on the sampled timestamp and not just if they have the same original timestamp. AFAIK one-to-one binary joins would explode like this. If we do this then we can discard the OrigT change which would make this PR a bit leaner. Would that be acceptable?

Reply:
Note: yeah this is important, we need to compare the actual TSDB sample timestamps so that the later info series can win. Its in the Info function spec., please ignore.

Reply:
Yep :)